src/templates/algorithms.h

```diff
@@ -94,3 +96,52 @@ template <class Lam> auto for_each_pair_early(int begin, int end, Lam lambda) ->
 
     return std::nullopt;
 }
+
+template <typename... Args> class ZipIterator
+{
+    public:
+    ZipIterator(std::tuple<Args...> args) : _source(args){};
+    bool operator!=(ZipIterator<Args...> other)
+    {
+        return std::apply(
+            [&other](auto &a, auto &... as) {
+                return std::apply(
+                    [&a](auto &b, auto &... bs) {
+                        // Only test the first elements.  We have to make the
+                        // assumption that all the containers are the same length,
+                        // anyway and this saves us some tests and some code.
+                        return a != b;
+                    },
+                    other._source);
+            },
+            _source);
+    }
+    void operator++()
+    {
+        std::apply([](auto &... item) { (item++, ...); }, _source);
+    }
+    auto operator*()
+    {
+        return std::apply([](auto &... item) { return std::make_tuple(std::ref(*item)...); }, _source);
+    }
+
+    private:
+    std::tuple<Args...> _source;
+};
+
+template <typename... Args> class zip
+{
+    public:
+    zip(Args &... args) : sources(args...) {}
+    auto begin()
+    {
+        return ZipIterator(std::apply([](auto &... item) { return std::make_tuple(item.begin()...); }, sources));
+    }
+    auto end()
+    {
+        return ZipIterator(std::apply([](auto &... item) { return std::make_tuple(item.end()...); }, sources));
+    }
+
+    private:
+    std::tuple<Args &...> sources;
```

Comment:
```suggestion
    std::tuple<Args &...> sources_;
```
For consistency with member visibility elsewhere.